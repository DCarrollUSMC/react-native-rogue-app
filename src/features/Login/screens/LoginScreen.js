/*
 * FILENAME:        features/Login/screens/LoginScreen.js
 *
 * DESCRIPTION:     Primary screen for Login feature.
 *
 * Author: SixPack Tech Team
 * Copyright (c) 2019 - Six-Pack Labs
 */

// React
import React, { Component } from 'react';
import { ImageBackground, TouchableOpacity, View, Image, KeyboardAvoidingView, Text, StatusBar } from 'react-native';

// Components
import SingleInput from '../../../components/SingleInput/SingleInput';
import ButtonSubmit from '../../../components/ButtonSubmit/ButtonSubmit';
import NetworkIssue from '../../../components/NetworkIssue/NetworkIssue';

// Import helper functions
import login from '../../../api/login';
import { defined, isConnected } from '../../../app/helpers';
import { appStorage } from '../../../app/storage';

// Validation
import validate from '../../../app/validation_wrapper';

// Styles and assets
import styles from '../styles/styles';
import usernameImg from '../../../assets/img/login/username.png';
import passwordImg from '../../../assets/img/login/password.png';
import kettleLoginImg from '../../../assets/img/login/kettle-login.jpg';
import spsImg from '../../../assets/img/login/sps-logo.png';

export default class LoginScreen extends Component {

    /**
     * Constructor
     *
     * @param obj props
     */
    constructor(props) {
        super(props);

        // Init blank state
        this.state = {
            email: '',
            password: '',
            emailError: '',
            message: null,
            isLoading: false,
            connInfo: null
        };

        // Bind input data
        this.handleEmailChange    = this.handleEmailChange.bind(this);
        this.handlePasswordChange = this.handlePasswordChange.bind(this);
        this.handleFormSubmit     = this.handleFormSubmit.bind(this);
    }

    async componentDidMount() {
        this.setState({ connInfo: await isConnected() })
    }

    /**
     * Mutate and set username/email to state lowercase
     *
     * @param string user
     */
    handleEmailChange(email) {
        this.setState({ email: email.toLowerCase() });
    }

    /**
     * Set password to state
     *
     * @param string pass
     */
    handlePasswordChange(pass) {
        this.setState({ password: pass });
    }

    /**
     * Action method that will trigger the HTTP request to AV2 API
     */
    async handleFormSubmit() {
        try {
            // Form validation
            this.setState({ emailError: validate('email', this.state.email) })

            if (!defined(this.state.emailError)) {
                this.setState({ isLoading: true })
                // Format outgoing data
                let res = await login({
                    email: this.state.email,
                    password: this.state.password
                });

                // Did not receive a response form login.js
                if (!defined(res)) {
                    // Make sure to add some behavior if we get back nothing.
                    // Ex: Some sort of messaging
                    this.setState({ message: 'Something went wrong!', isLoading: false });
                }

                // Reset the form
                this.clearForm();

                // Check the returned response
                switch (res.response) {
                    case 'success':
                        // Successful authentication, set state and redirect to dashboard
                        appStorage.setItem('USER_TOKEN', res.result.token);
                        this.props.navigation.navigate('App');
                        break;
                    case 'error':
                        this.setState({ message: res.message, isLoading: false });
                        break;
                }
            }
        } catch(err) {
            console.log(err);
        }
    }

    _onPress() {
        this.props.navigation.navigate('ForgotPass');
    }

    clearForm() {
        this.setState({
            email: '',
            password: ''
        });
    }

    /**
     * Render component
     */
    render(){
        if(this.state.connInfo === false) {
            return <NetworkIssue />
        }
        return(
            <ImageBackground source={kettleLoginImg} imageStyle={{}} style={styles.loginBG}>
                <StatusBar barStyle='light-content' />
                <KeyboardAvoidingView behavior="padding" style={styles.container}>
                    <Image style={styles.logo} source={spsImg}/>
                    {defined(this.state.message) &&
                        <Text style={styles.error}>{this.state.message}</Text>
                    }
                    <SingleInput
                        source={usernameImg}
                        placeholder="Email"
                        autoCapitalize={'none'}
                        returnKeyType={'done'}
                        autoCorrect={false}
                        onChangeText={this.handleEmailChange}
                        value={this.state.email}
                        iconWidth={22}
                        iconHeight={22}
                        placeholderTextColor='#FFF'
                        bgColor='#333'
                        txtColor='#FFF'
                        underlineColorAndroid='transparent'
                        borderBottomColor='transparent'
                        error={this.state.emailError}
                    />
                    <SingleInput
                        source={passwordImg}
                        secureTextEntry={true}
                        placeholder="Password"
                        returnKeyType={'done'}
                        autoCapitalize={'none'}
                        autoCorrect={false}
                        onChangeText={this.handlePasswordChange}
                        value={this.state.password}
                        iconWidth={22}
                        iconHeight={22}
                        placeholderTextColor='#FFF'
                        bgColor='#333'
                        txtColor='#FFF'
                        underlineColorAndroid='transparent'
                        borderBottomColor='transparent'
                    />
                    <ButtonSubmit formSubmit={this.handleFormSubmit} label="Login" isLoading={this.state.isLoading}/>
                    <View style={styles.forgotPass}>
                        <Text style={[styles.text, styles.forgotPassLink]}>
                            Forgot Password?
                        </Text>
                        <TouchableOpacity onPress={() => this.props.navigation.navigate('ForgotPassword')}>
                            <Text style={styles.mainlink}>
                                Recover here
                            </Text>
                        </TouchableOpacity>
                    </View>
                </KeyboardAvoidingView>
            </ImageBackground>
        )
    }
}