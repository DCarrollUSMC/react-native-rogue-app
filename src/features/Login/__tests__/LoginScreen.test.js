/*
 * FILENAME:        features/Login/__tests__/LoginScreen.test.js
 *
 * DESCRIPTION:     Test file for LoginScreen Component
 * 
 * Author: SixPack Tech Team
 * Copyright (c) 2019 - Six-Pack Labs
 */

import React from 'react';
import renderer from 'react-test-renderer';

import LoginScreen from '../screens/LoginScreen';

// Mock SingleInput component to get LoginScreen to render in jest (ignoring SingleInput)
jest.mock('../../../components/SingleInput/SingleInput', () => 'SingleInput');

describe('LoginScreen', () => {
    it('renders correctly', () => {
        const tree = renderer.create(
            <LoginScreen />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});