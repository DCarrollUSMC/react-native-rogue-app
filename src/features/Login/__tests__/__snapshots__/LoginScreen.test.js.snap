// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/img/login/kettle-login.jpg",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        Object {},
      ]
    }
  />
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(34, 34, 34, 0.95)",
          "flex": 1,
          "justifyContent": "center",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/img/login/sps-logo.png",
        }
      }
      style={
        Object {
          "bottom": 60,
          "height": 60,
          "resizeMode": "contain",
          "width": 60,
        }
      }
    />
    <SingleInput
      autoCapitalize="none"
      autoCorrect={false}
      bgColor="#333"
      borderBottomColor="transparent"
      iconHeight={22}
      iconWidth={22}
      onChangeText={[Function]}
      placeholder="Email"
      placeholderTextColor="#FFF"
      returnKeyType="done"
      source={
        Object {
          "testUri": "../../../src/assets/img/login/username.png",
        }
      }
      txtColor="#FFF"
      underlineColorAndroid="transparent"
      value=""
    />
    <SingleInput
      autoCapitalize="none"
      autoCorrect={false}
      bgColor="#333"
      borderBottomColor="transparent"
      iconHeight={22}
      iconWidth={22}
      onChangeText={[Function]}
      placeholder="Password"
      placeholderTextColor="#FFF"
      returnKeyType="done"
      secureTextEntry={true}
      source={
        Object {
          "testUri": "../../../src/assets/img/login/password.png",
        }
      }
      txtColor="#FFF"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "flex-start",
          "margin": 10,
          "top": 0,
        }
      }
    >
      <View
        style={
          Object {
            "width": 710,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <BVLinearGradient
            colors={
              Array [
                4282553121,
                4280849547,
              ]
            }
            endPoint={
              Object {
                "x": 1,
                "y": 0,
              }
            }
            locations={null}
            startPoint={
              Object {
                "x": 0,
                "y": 0,
              }
            }
            style={
              Object {
                "borderRadius": 5,
                "height": 50,
                "width": 710,
              }
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "transparent",
                  "color": "white",
                  "fontSize": 18,
                  "fontWeight": "800",
                  "letterSpacing": 2,
                  "margin": 10,
                  "textAlign": "center",
                }
              }
            >
              LOGIN
            </Text>
          </BVLinearGradient>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFF",
            },
            Object {
              "paddingRight": 3,
            },
          ]
        }
      >
        Forgot Password?
      </Text>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#429321",
            }
          }
        >
          Recover here
        </Text>
      </View>
    </View>
  </View>
</View>
`;
